syntax = "proto3";

package gtools.growtopia;

message State {
  World world = 1;
  Me me = 3;
  uint32 status = 4;
  Inventory inventory = 5;
}

message Inventory { repeated Item items = 1; }

message CharacterState {
  uint32 net_id = 9;
  uint32 build_range = 1;
  uint32 punch_range = 2;
  uint32 flags = 3;
  float gravity = 4;
  float velocity = 5;
  float acceleration = 6;
  float velocity_in_water = 7;
  float jump_strength = 8;
}

message Me {
  int32 net_id = 3;
  uint32 build_range = 1;
  uint32 punch_range = 2;
  Vec2F pos = 4;
  uint32 flags = 9;
  CharacterState state = 10;
  uint32 server_ping = 5;
  uint32 client_ping = 8;
  float time_since_login = 6;
  float time_in_world = 7;
}

message Item {
  uint32 id = 1;
  uint32 amount = 2;
  uint32 flags = 3;
}

message Vec2I {
  uint32 x = 1;
  uint32 y = 2;
}

message Vec4I {
  uint32 x = 1;
  uint32 y = 2;
  uint32 w = 3;
  uint32 h = 4;
}

message Vec2F {
  float x = 1;
  float y = 2;
}

message Player {
  bytes spawn = 1;   // avatar
  uint32 netID = 2;  // 1
  uint32 userID = 3; // 12839123
  bytes eid = 4;     // 12839123  awojdioajwidj=== | aowdjoiawjd===
  bytes ip = 5;      // aoiwjdoj1i2j3o123= (b64decode length 32)
  Vec4I colrect = 6; // 0 0 20 20
  Vec2F posXY = 7;   // 1820 810
  bytes name = 8;
  bytes titleIcon =
      9; // {"PlayerWorldID":1,"WrenchCustomization":{"WrenchForegroundCanRotate":false,"WrenchForegroundID":-1,"WrenchIconID":-1}}
         // absolute insanity
  bytes country = 10;  // us | showGuild
  uint32 invis = 11;   // 0
  uint32 mstate = 12;  // 0
  uint32 smstate = 13; // 0
  bytes onlineID = 14;

  uint32 flags = 15;
  CharacterState state = 16;
}

message DroppedItem {
  uint32 id = 1;
  float x = 2;
  float y = 3;
  uint32 amount = 4;
  uint32 flags = 5;
  uint32 uid = 6;
}

message Dropped {
  uint32 nb_items = 1;
  uint32 last_uid = 2;
  repeated DroppedItem items = 3;
}

message Tile {
  uint32 fg_id = 1;
  uint32 bg_id = 2;
  uint32 lock_index = 3;
  uint32 parent_index = 11;
  uint32 flags = 4;
  bytes extra = 5;
  uint32 index = 10;
  uint32 x = 6;
  uint32 y = 7;
  bytes json_data = 9;
}

// the world data
message WorldInner {
  bytes name = 1;
  uint32 width = 5;
  uint32 height = 6;
  uint32 nb_tiles = 7;
  repeated Tile tiles = 2;
  Dropped dropped = 4;
  int32 garbage_start = 8;
}

// the world state (not included in data)
message World {
  WorldInner inner = 1;
  repeated Player player = 3;
  repeated Npc npcs = 2;
}

message Npc {
  uint32 id = 1;
  uint32 state = 2;
  float x = 3;
  float y = 4;
  float target_x = 5;
  float target_y = 6;
  int32 param1 = 7;
  int32 param2 = 8;
  float param3 = 9;
  bool facing_left = 10;
}
